wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+from copy import deepcopy
+import logging
+
+import numpy as np
+from sklearn.manifold import TSNE
+from sklearn.model_selection import ParameterGrid
+from sklearn.metrics import silhouette_score
+from sklearn.cluster import DBSCAN
+import umap
+
+from wellcomeml.ml import Vectorizer
+from wellcomeml.logger import logger
+
+
+class TextClustering(object):
+    """
+    Class
+
+    Attributes:
+        vectorizer: The embedding Vectorizer object
+        reducer: A dimensionality reduction object
+        clustering: A clustering model object
+        cluster_ids: Ids of the cluster
+        cluster_names: Names of the clusters
+        cluster_kws: Keywords for the clusters (only if embedding=tf-idf)
+    """
+    def __init__(self, embedding='tf-idf', reducer='umap', clustering='dbscan',
+                 n_kw=10, params={}):
+        """
+        Initialises a cluster pipeline
+
+        Args:
+            embedding(str): 'tf-idf', 'bert', 'keras' or 'doc2vec'.
+            reducer(str): 'umap' or 'tsne'
+            clustering(str): Only 'dbscan' allowed for now
+            n_kw(int): If n_kw > 0 and embedding='tf-idf', the model
+            outputs, besides a cluster id, a list of keyword for each
+            cluster.
+            params(dict): Dictionary containing extra parameters for
+            embedding, reducer or clustering models. Example:
+            {'clustering': {'eps': 0.1}, 'embedding': {'ngram_range': (1,2)}}
+        """
+        self.embedding = embedding
+        self.vectorizer = Vectorizer(embedding=embedding,
+                                     **params.get('embedding', {}))
+        self.n_kw = n_kw
+
+        reducer_dispatcher = {
+            'umap': umap.UMAP,
+            'tsne': TSNE
+        }
+        self.reducer = reducer_dispatcher[reducer](**params.get('reducer', {}))
+
+        clustering_dispatcher = {
+            'dbscan': DBSCAN
+        }
+
+        self.clustering = clustering_dispatcher[clustering](
+            **params.get('clustering', {})
+        )
+
+        self.cluster_ids = None
+        self.cluster_names = None
+        self.cluster_kws = None
+
+    def fit(self, X, *_):
+        """
+        Fits all clusters in the pipeline
+
+        Args:
+            X: A list of texts
+
+        Returns:
+            A TextClustering object
+
+        """
+        self._fit_step(X, step='vectorizer')
+        self._fit_step(step='reducer')
+        self._fit_step(step='clustering')
```

Comment:
so you first reduce and then cluster. i wonder whether that removes the ability to generalise to other clustering techniques such as lda, nmf, k means etc. see also here https://scikit-learn.org/stable/auto_examples/applications/plot_topics_extraction_with_nmf_lda.html#sphx-glr-auto-examples-applications-plot-topics-extraction-with-nmf-lda-py. i think k means is propably fine. nmf works but does it work as expected? lda i think only makes sense before reducing

Reply:
I agree that LDA only makes sense if you have the raw vectors. For uMAP specifically you often reduce to an intermediate dimension for clustering and a further reduction to 2 for visualisation, although the intermediate reduction did not help my case.

For the time being, I added a possibility of skipping reduction by a parameter `cluster_reduced` in https://github.com/wellcometrust/WellcomeML/pull/109/commits/b0910f00b612a853a9073c3b7d0c09ccb04208b3